### PR TITLE
Implement kitty breeder functions 

### PR DESCRIPTION
**Pull request by PaulsBecks**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -178,24 +179,158 @@ func isPregnant(kittyID uint64) (bool, error) {
...
+		return false, nil
+	}
+
+	matron := kitties[matronID]
+	sire := kitties[sireID]
+
+	if matron.MatronID == sireID || matron.SireID == sireID {
+		return false, nil
+	}
+	if sire.MatronID == sireID || sire.SireID == sireID {
```
Comment (rohrschacht) [inline main.go]: I am comparing the solidity implementation with this side-by-side. In this line, the ID to compare against should be matronID both times (eg sire.MatronID == matronID)

File: main.go [hunk 2]
```diff
@@ -178,24 +179,158 @@ func isPregnant(kittyID uint64) (bool, error) {
...
+	if err != nil {
+		return err
+	}
+
+	matronOwner := kittyIndexToOwner[matronID]
+
+	if matronOwner != clientID {
+		return fmt.Errorf("Caller must be the owner of the matron kitty.")
+	}
+
```
Comment (rohrschacht) [inline main.go:285]: You have checked at this point that the matron is owned by the message sender. After that, in the solidity implementation, there is a requirement for isSiringPermitted, eg:
```solidity
require(_isSiringPermitted(_sireId, _matronId));
```
This is after the owner is checked and before the ReadyToBreed functions are called. I think this isSiringPermitted() is missing here.

File: main.go [hunk 3]
```diff
@@ -178,24 +179,158 @@ func isPregnant(kittyID uint64) (bool, error) {
...
+		return fmt.Errorf("Matron with id %d and sire with id %d are no valid mating pair. Shame on you! Don't try to breed these cats.", matronID, sireID)
+	}
+
+	return breedWith(sireID, matronID)
 }
 
-func (c *KittyContract) GiveBirth(ctx contractapi.TransactionContextInterface, matronID uint64) error {
-	return nil
+func mixGenes(matronGenes, sireGenes uint64) uint64 {
+	return matronGenes ^ sireGenes // ^ will implement xor
```
Comment (rohrschacht) [inline main.go:306]: xor is a very nice idea for the mixing! :D :+1:

File: main.go [hunk 4]
```diff
@@ -178,24 +179,158 @@ func isPregnant(kittyID uint64) (bool, error) {
...
+		return 0, fmt.Errorf("No kitty with this ID available.")
+	}
+
+	matron := kitties[matronID]
+
+	if ok, err := isReadyToGiveBirth(matron); err != nil || !ok {
+		return 0, fmt.Errorf("Matron is not yet ready to give birth.")
+	}
+
+	sireID := matron.SireID
```
Comment (rohrschacht) [inline main.go]: In the solidity implementation, the kitty with the ID matron.SiringWithID is used here. This should be the correct one, as matron.SireID should be the sire (father) of the matron we are given.

Comment (rohrschacht) [reply main.go]: After fixing here which ID is grabbed from the matron, the rest of the code should logically work :+1:

**Review by rohrschacht — CHANGES_REQUESTED** [review]
I found 3 little logic errors that should be easily fixable. Otherwise the logic looks very good!

File: main.go (hunk 1 above)
Comment (PaulsBecks) [reply main.go]: ah true, of course. Thanks.

File: main.go (hunk 3 above)
Comment (PaulsBecks) [reply main.go:306]: :)

**Review by rohrschacht — APPROVED** [review]
I can't find any logic errors in this. I ran `golint` and `go vet` and nothing serious comes up. I say please merge it! :+1:

**Pull request merged**